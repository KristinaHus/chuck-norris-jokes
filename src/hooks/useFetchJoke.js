import { useQuery } from 'react-query';

import { getJoke } from '../services/jokes';
import { FETCH_DELAY } from '../config/constansts';

export function useFetchJoke(enabled) {
	return useQuery('joke', () => getJoke(), {
		refetchInterval: FETCH_DELAY,
		enabled
	});
}
