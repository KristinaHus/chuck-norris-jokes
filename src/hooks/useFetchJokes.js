import { useQuery } from 'react-query';
import { getJokes } from '../services/jokes';

export function useFetchJokes() {
	return useQuery('jokes', () => getJokes(), {
		refetchOnWindowFocus: false
	});
}
