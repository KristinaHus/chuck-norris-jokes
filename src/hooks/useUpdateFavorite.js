import { useIndexedDB } from 'react-indexed-db';
import { useEffect, useState } from 'react';

import { MAX_COMMENTS_AMOUNT } from '../config/constansts';

export function useUpdateFavorite(item) {
	const [isFavorite, setIsFavorite] = useState(false);
	const [error, setError] = useState('');
	const { add, deleteRecord, getByIndex, getAll } = useIndexedDB('favorite');

	useEffect(() => {
		getAllFavorites();
	}, []);

	const fetchFavorites = async () => {
		return getAll();
	};

	const getAllFavorites = async () => {
		const favorites = await fetchFavorites();
		if (
			favorites.findIndex((favorite) => {
				return favorite.id === item.id;
			}) > -1
		) {
			setIsFavorite(true);
		}
	};

	const updateFavorite = async () => {
		const favorites = await fetchFavorites();
		const isMaxReached = favorites.length >= MAX_COMMENTS_AMOUNT;
		if (error) {
			setError('');
		}

		setIsFavorite((favorite) => {
			if (!favorite) {
				if (isMaxReached) {
					setError(
						'You already have the maximum amount of favorite jokes. Remove one to be able to add a new one.'
					);
				} else {
					add({ id: item.id, value: item.value }).then(
						(id) => {
							console.log('ID Generated: ', id);
						},
						(error) => {
							console.log(error);
						}
					);
				}
			} else {
				getByIndex('id', item.id).then((dbRecord) => {
					deleteRecord(dbRecord._id);
				});
			}
			return isMaxReached ? false : !favorite;
		});
	};

	return { isFavorite, updateFavorite, error };
}
