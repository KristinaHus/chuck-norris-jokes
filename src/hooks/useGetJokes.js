import { useEffect, useState } from 'react';

import { removeLatest } from '../helpers/filters';
import { useFetchJoke } from './useFetchJoke';
import { useFetchJokes } from './useFetchJokes';
import { FETCH_DELAY } from '../config/constansts';

export function useGetJokes() {
	const [jokes, setJokes] = useState([]);
	const [isJokeFetchEnabled, setIsJokeFetchEnabled] = useState(false);

	const { data, isLoading, isError } = useFetchJokes();
	const { data: joke } = useFetchJoke(isJokeFetchEnabled);

	useEffect(() => {
		setTimeout(() => {
			setIsJokeFetchEnabled(true);
		}, FETCH_DELAY);
	}, []);

	useEffect(() => {
		if (data && data.length > 0 && !isError) {
			setJokes(data);
		}
	}, [data]);

	useEffect(() => {
		if (joke && jokes) {
			const jokesList = [joke, ...removeLatest(jokes)];
			setJokes(jokesList);
		}
	}, [joke]);

	return { jokes, isLoading };
}
