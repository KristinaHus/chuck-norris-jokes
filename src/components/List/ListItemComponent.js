import { useEffect, useState } from 'react';
import {
	ListItem,
	ListItemButton,
	ListItemIcon,
	ListItemText,
	Snackbar,
	Alert
} from '@mui/material';
import FavoriteIcon from '@mui/icons-material/Favorite';
import FavoriteBorderIcon from '@mui/icons-material/FavoriteBorder';
import PropTypes from 'prop-types';

import { useUpdateFavorite } from '../../hooks/useUpdateFavorite';

export const ListItemComponent = ({ item, onPress }) => {
	const [isError, setIsError] = useState(false);
	const { isFavorite, updateFavorite, error } = useUpdateFavorite(item);

	useEffect(() => {
		setIsError(!!error);
	}, [error]);

	const handleClose = () => {
		setIsError(false);
	};

	const handleClick = () => {
		updateFavorite();
		setIsError(!!error);
		if (onPress) {
			setTimeout(() => {
				onPress();
			}, 500);
		}
	};

	return (
		<ListItem disablePadding data-testid="ListItem">
			<ListItemButton onClick={handleClick} data-testid="favoriteButton">
				<ListItemIcon>{isFavorite ? <FavoriteIcon /> : <FavoriteBorderIcon />}</ListItemIcon>
				<ListItemText primary={item.value} />
			</ListItemButton>
			<Snackbar open={isError} autoHideDuration={1000 * 3} onClose={handleClose}>
				<Alert onClose={handleClose} severity="error" sx={{ width: '100%' }}>
					{error}
				</Alert>
			</Snackbar>
		</ListItem>
	);
};

ListItemComponent.propTypes = {
	item: PropTypes.object,
	onPress: PropTypes.func
};
