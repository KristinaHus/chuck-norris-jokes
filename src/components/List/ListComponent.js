import { List } from '@mui/material';
import { ListItemComponent } from './ListItemComponent';

export const ListComponent = () => {
	const data = [
		{ id: 'one', value: 'one' },
		{ id: 'two', value: 'two' },
		{ id: 'three', value: 'three' },
		{ id: 'four', value: 'four' },
		{ id: 'five', value: 'five' },
		{ id: 'six', value: 'six' },
		{ id: 'seven', value: 'seven' },
		{ id: 'eight', value: 'eight' },
		{ id: 'nine', value: 'nine' },
		{ id: 'ten', value: 'ten' },
		{ id: 'eleven', value: 'eleven' },
		{ id: 'twelve', value: 'twelve' }
	];

	return (
		<div>
			<List
				sx={{ width: '100%', maxWidth: 360, bgcolor: 'background.paper' }}
				aria-label="contacts">
				{data.map((item) => (
					<ListItemComponent item={item} key={item.id} />
				))}
			</List>
		</div>
	);
};
