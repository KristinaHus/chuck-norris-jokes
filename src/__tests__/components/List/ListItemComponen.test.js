import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';

import { ListItemComponent } from '../../../components/List/ListItemComponent';
import { useUpdateFavorite } from '../../../hooks/useUpdateFavorite';

jest.mock('../../../hooks/useUpdateFavorite');

describe('Test ListComponent icon change', () => {
	it('List item is not marked as favorite', async () => {
		useUpdateFavorite.mockReturnValue({ isFavorite: false, updateFavorite: () => {}, err: '' });
		const Component = () => (
			<ListItemComponent
				item={{
					id: '1',
					value: 'Test value'
				}}
			/>
		);

		render(<Component />);

		// Test snapshots
		const component = renderer.create(<Component />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();

		expect(screen.findByTestId('favoriteButton')).toBeTruthy();
		expect(screen.queryByTestId('FavoriteBorderIcon')).toBeInTheDocument();
		expect(screen.queryByTestId('FavoriteIcon')).not.toBeInTheDocument();
	});

	it('List item is marked as favorite', async () => {
		useUpdateFavorite.mockReturnValue({ isFavorite: true, updateFavorite: () => {}, err: '' });

		const Component = () => (
			<ListItemComponent
				item={{
					id: '1',
					value: 'Test value'
				}}
			/>
		);

		render(<Component />);
		const component = renderer.create(<Component />);

		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();

		expect(screen.queryByTestId('favoriteButton')).toBeInTheDocument();
		expect(screen.queryByTestId('FavoriteIcon')).toBeInTheDocument();
		expect(screen.queryByTestId('FavoriteBorderIcon')).not.toBeInTheDocument();
	});
});
