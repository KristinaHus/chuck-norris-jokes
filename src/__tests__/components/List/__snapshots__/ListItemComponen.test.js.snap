// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ListComponent icon change List item is marked as favorite 1`] = `
<li
  className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
  data-testid="ListItem"
  disabled={false}
>
  <div
    className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
    data-testid="favoriteButton"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="FavoriteIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
        />
      </svg>
    </div>
    <div
      className="MuiListItemText-root css-tlelie-MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
      >
        Test value
      </span>
    </div>
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </div>
</li>
`;

exports[`Test ListComponent icon change List item is not marked as favorite 1`] = `
<li
  className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
  data-testid="ListItem"
  disabled={false}
>
  <div
    className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
    data-testid="favoriteButton"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="FavoriteBorderIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
        />
      </svg>
    </div>
    <div
      className="MuiListItemText-root css-tlelie-MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
      >
        Test value
      </span>
    </div>
  </div>
</li>
`;
